src/protocols/screencopy.rs

```diff
@@ -275,44 +287,70 @@ where
 
         let (buffer, with_damage) = match request {
             zwlr_screencopy_frame_v1::Request::Copy { buffer } => (buffer, false),
-            // zwlr_screencopy_frame_v1::Request::CopyWithDamage { buffer } => (buffer, true),
+            zwlr_screencopy_frame_v1::Request::CopyWithDamage { buffer } => (buffer, true),
             _ => unreachable!(),
         };
 
-        if !shm::with_buffer_contents(&buffer, |_buf, shm_len, buffer_data| {
+        let buffer = if let Ok(dmabuf) = dmabuf::get_dmabuf(&buffer) {
+            if !(dmabuf.format().code == Fourcc::Argb8888
+                && dmabuf.size().w == info.buffer_size.w
+                && dmabuf.size().h == info.buffer_size.h)
+            {
+                frame.post_error(
+                    zwlr_screencopy_frame_v1::Error::InvalidBuffer,
+                    "invalid buffer",
+                );
+                return;
+            }
+            ScreencopyBuffer::Dmabuf(dmabuf.clone())
+        } else if shm::with_buffer_contents(&buffer, |_buf, shm_len, buffer_data| {
             buffer_data.format == wl_shm::Format::Argb8888
                 && buffer_data.stride == info.buffer_size.w * 4
                 && buffer_data.height == info.buffer_size.h
                 && shm_len as i32 == buffer_data.stride * buffer_data.height
         })
         .unwrap_or(false)
         {
+            ScreencopyBuffer::Shm(buffer)
+        } else {
             frame.post_error(
                 zwlr_screencopy_frame_v1::Error::InvalidBuffer,
                 "invalid buffer",
             );
             return;
-        }
+        };
 
         copied.store(true, Ordering::SeqCst);
 
-        state.frame(Screencopy {
-            with_damage,
+        let damages = state.frame(Screencopy {
```

Comment:
So, the semantics for copy_with_damage() is a bit different from copy() (which is one of the reasons why I haven't implemented v3 initially). Whereas copy() can always copy the current frame immediately, as it does here, copy_with_damage() cannot:

> Same as copy, except it waits until there is damage to copy.

As far as I understand this PR, if a client issues copy_with_damage() before there's been any damage, niri will copy the frame and submit it right away, and not submit any damage. This is wrong, as far as I can tell from the quoted description from the protocol.

Instead, what should happen, is that every time the corresponding output is rendered, it should also get checked against the screencopy damage tracker, and if there's any damage, then it should get submitted to screencopy (damage event goes first).